### PR TITLE
Combine Support: Sign in with 3rd party credentials

### PR DESCRIPTION
**Pull request by ziadtamim**
Proposal for Combine support to authenticate with Firebase using third-party credentials.

```swift 
   func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error?) {
       // ...
       if let error = error {
         // ...
         return
       }

       guard let authentication = user.authentication else { return }
       let credential = GoogleAuthProvider.credential(withIDToken: authentication.idToken,
                                                            accessToken: authentication.accessToken)
        Auth.auth()
            .signIn(withCredential: credential)
            .mapError { $0 as NSError }
            .tryCatch(handleError)
            .sink { /* ... */ } receiveValue: {  /* ... */  }
            .store(in: &subscriptions)
    }
    
    private func handleError(_ error: NSError) throws -> AnyPublisher<AuthDataResult, Error> {
        guard isMFAEnabled && error.code == AuthErrorCode.secondFactorRequired.rawValue
        else { throw error }
        
        // The user is a multi-factor user. Second factor challenge is required.
        let resolver = error.userInfo[AuthErrorUserInfoMultiFactorResolverKey] as! MultiFactorResolver
        let displayNameString = resolver.hints.compactMap(\.displayName).joined(separator: " ")
        
        return showTextInputPrompt(withMessage: "Select factor to sign in\n\(displayNameString)")
            .compactMap { displayName in
                resolver.hints.first(where: { displayName == $0.displayName }) as? PhoneMultiFactorInfo
            }
            .flatMap { [unowned self] factorInfo in
                PhoneAuthProvider.provider()
                    .verifyPhoneNumber(withMultiFactorInfo: factorInfo, multiFactorSession: resolver.session)
                    .zip(self.showTextInputPrompt(withMessage: "Verification code for \(factorInfo.displayName ?? "")"))
                    .map { (verificationID, verificationCode) in
                        let credential = PhoneAuthProvider.provider().credential(withVerificationID: verificationID, 
                                                                                 verificationCode: verificationCode)
                        return PhoneMultiFactorGenerator.assertion(with: credential)
                    }
            }.flatMap { assertion in
                resolver.resolveSignIn(withAssertion: assertion)
            }.eraseToAnyPublisher()
    }
```

### CODE REVIEW

**Comment by peterfriese** [comment]
HI @ziadtamim - thanks for your contribution!

I've just merged #6976 into `combine-main`, laying the groundwork for Combine support for Firebase Auth. Can you update this PR and:

1. Make sure it targets `combine-main`
2. Add some tests for your extensions. [FirebaseAuth/Tests/Unit](https://github.com/firebase/firebase-ios-sdk/tree/master/FirebaseAuth/Tests/Unit) contains tests for all the Auth providers, and you can take some inspiration for how to write the respective mocks from [FirebaseCombineSwift/Tests/Unit/Auth/EmailLinkAuthTests.swift](https://github.com/firebase/firebase-ios-sdk/blob/d90594069af8f18eeaa133d9d14ab482bd510131/FirebaseCombineSwift/Tests/Unit/Auth/EmailLinkAuthTests.swift)

**Comment by peterfriese** [comment]
Ooops - looks like you'll first need to catch up to the head of combine-main before attempting the merge.

**Comment by ziadtamim** [comment]
> Ooops - looks like you'll first need to catch up to the head of combine-main before attempting the merge. There is a lot of noise in your most recent change that makes it next to impossible to see the actual change you're proposing:
> 
> Screenshot 2021-01-15 at 09 22 55
> 
> Alternatively, create a new PR agains `combine-main`, and then abandon this one. This might actually be the easist way forward.

Thanks @peterfriese I'll fix the PR and add tests.

**Comment by ziadtamim** [comment]
@peterfriese PR fixed with added tests.

**Review by peterfriese — CHANGES_REQUESTED** [review]
Looking good. 

Some minor naming changes in the new methods you added - see my detailed comments there.

Thanks for adding the tests! Had one minor issue there as well.

**Review by peterfriese — COMMENTED** [review]
Thanks for making those changes, @ziadtamim!

If you find some time, can you also add a short example to the README file in the root of FirebaseCombine?

**Comment by ziadtamim** [comment]
@peterfriese README update ✔

File: FirebaseCombineSwift/README.md [hunk 1]
```diff
@@ -58,4 +58,54 @@ end
     .replaceError(with: "(unable to sign in anonymously)")
     .assign(to: \.uid, on: self)
     .store(in: &cancellables)
-```
\ No newline at end of file
+```
+#### Sign in with a given 3rd-party credentials 
+
+
+In the `sign(_:didSignInFor:withError:)` method, get a Google ID token and Google access token from the GIDAuthentication object and exchange them for a Firebase credential. Then, Asynchronously sign in to Firebase with the given tokens pair:
```
Comment (peterfriese) [inline FirebaseCombineSwift/README.md:65]: I think the official docs say it quite nicely: 

> In the `signIn:didSignInForUser:withError:` method, get a Google ID token and Google access token from the `GIDAuthentication` object and exchange them for a Firebase credential:

Let's use this and change your paragraph to:

> In the `sign(_:didSignInFor:withError:)` method, get a Google ID token and Google access token from the `GIDAuthentication` object and asynchronously exchange them for a Firebase credential:

File: FirebaseCombineSwift/README.md [hunk 2]
```diff
@@ -58,4 +58,54 @@ end
...
+```swift 
+    func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error?) {
+       // ...
+       if let error = error {
+         // ...
+         return
+       }
+
+       guard let authentication = user.authentication else { return }
+       let credential = GoogleAuthProvider.credential(withIDToken: authentication.idToken,
```
Comment (peterfriese) [inline FirebaseCombineSwift/README.md:76]: Please make sure to align the `withIDToken` and `withIDToken`. (I am painfully aware our own docs often aren't a good example ...)

File: FirebaseCombineSwift/README.md [hunk 3]
```diff
@@ -58,4 +58,54 @@ end
...
\ No newline at end of file
+```
+#### Sign in with a given 3rd-party credentials 
+
+
+In the `sign(_:didSignInFor:withError:)` method, get a Google ID token and Google access token from the GIDAuthentication object and exchange them for a Firebase credential. Then, Asynchronously sign in to Firebase with the given tokens pair:
+
+```swift 
+    func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error?) {
+       // ...
```
Comment (peterfriese) [inline FirebaseCombineSwift/README.md:69]: Google indentation style is 2 spaces. Can you update the entire snippet accordingly?

File: FirebaseCombineSwift/README.md [hunk 4]
```diff
@@ -58,4 +58,54 @@ end
...
+            .flatMap { [unowned self] factorInfo in
+                PhoneAuthProvider.provider()
+                    .verifyPhoneNumber(withMultiFactorInfo: factorInfo, multiFactorSession: resolver.session)
+                    .zip(self.showTextInputPrompt(withMessage: "Verification code for \(factorInfo.displayName ?? "")"))
+                    .map { (verificationID, verificationCode) in
+                        let credential = PhoneAuthProvider.provider().credential(withVerificationID: verificationID, 
+                                                                                 verificationCode: verificationCode)
+                        return PhoneMultiFactorGenerator.assertion(with: credential)
+                    }
+            }.flatMap { assertion in
```
Comment (peterfriese) [inline FirebaseCombineSwift/README.md:107]: Bring .flatMap onto the next line, just like you did in line 98.

File: FirebaseCombineSwift/README.md [hunk 5]
```diff
@@ -58,4 +58,54 @@ end
...
+                resolver.resolveSignIn(withAssertion: assertion)
+            }.eraseToAnyPublisher()
```
Comment (peterfriese) [inline FirebaseCombineSwift/README.md:109]: Same here - this way, all calls are nicely aligned.

**Review by peterfriese — CHANGES_REQUESTED** [review]
Looking good. Just some formatting issues. Once you've addressed them, we can merge this PR.

**Comment by ziadtamim** [comment]
Requested changes applied.

**Comment by peterfriese** [comment]
Awesome, thanks for making those changes.

**Review by peterfriese — COMMENTED** [review]
LGTM!

**Pull request merged**